Typing into Facebook post autocompletes too much

rr-codes opened the pull request:
#### 09f5cc35881121c7761675ecf6ade4eb51ef8fb5
<pre>
Typing into Facebook post autocompletes too much
<a href="https://bugs.webkit.org/show_bug.cgi?id=276206">https://bugs.webkit.org/show_bug.cgi?id=276206</a>
<a href="https://rdar.apple.com/129435650">rdar://129435650</a>

Reviewed by Wenson Hsieh.

Use the new pseudo-style-based writing suggestions implementation on both iOS and macOS, instead of just macOS,
and account for the differences in the behavior between how AppKit and UIKit interface with WebKit for providing
the predictive text.

* Source/WebCore/dom/WritingSuggestionData.h:
(WebCore::WritingSuggestionData::WritingSuggestionData):
(WebCore::WritingSuggestionData::originalPrefix const):
(WebCore::WritingSuggestionData::originalSuffix const):
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::setWritingSuggestion):
(WebCore::Editor::respondToChangedSelection):
* Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
(WebCore::RenderTreeUpdater::GeneratedContent::updateWritingSuggestionsRenderer):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView setAttributedMarkedText:selectedRange:]):

Canonical link: <a href="https://commits.webkit.org/281060@main">https://commits.webkit.org/281060@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ff107ba37fa0c0aeac5d7dfee9a573b87d689310

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58556 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37883 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11041 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62182 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9000 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/60685 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45519 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9198 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47405 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6414 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60587 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35462 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50628 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28257 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32210 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7930 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8004 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54163 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8202 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63885 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2469 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8211 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54726 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2477 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50654 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54808 "Found 1 new API test failure: /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print-errors (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12933 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2088 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33712 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34798 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35882 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34543 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -6296,6 +6296,24 @@ static BOOL shouldUseHighlightsForMarkedText(NSAttributedString *string)
+    BOOL hasTextCompletion = ^{
+        // UIKit doesn't include the `NSTextCompletionAttributeName`, so the next best way to detect if this method
+        // is being used for a text completion is to check if the attributes match these hard-coded ones.
+        RetainPtr textCompletionAttributes = @{
```
Hm... this is a bit of a hack, but I don't really have a cleaner suggestion in the moment :/

Perhaps we could adjust this to return `YES`, iff. the `markedTextAttributes` _contains_ both a `NSForegroundColorAttributeName` of `systemGrayColor` and `NSBackgroundColorAttributeName` of `clearColor`?

That way, if we clean this up in the future by specifying an new key in this dictionary like `NSTextCompletionAttributeName`, doing so won't regress WebKit and we can then safely remove this logic (in favor of checking for `NSTextCompletionAttributeName`, or the future UIKit equivalent)

pxlcoder commented on Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
```diff
@@ -318,9 +318,15 @@ void RenderTreeUpdater::GeneratedContent::updateWritingSuggestionsRenderer(Rende
 
     auto textWithoutSuggestion = nodeBeforeWritingSuggestionsTextRenderer->text();
 
+#if PLATFORM(MAC)
```
If iOS is the special case, I think we should use that rather than `PLATFORM(MAC)`.

whsieh commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -3987,6 +3997,43 @@ void Editor::respondToChangedSelection(const VisibleSelection&, OptionSet<FrameS
+        if (!startsWith)
+            return false;
+
+        return startsWith;
```
This is probably cleaner:

```suggestion
        return completion.startsWith(content);
```

whsieh commented on Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
```diff
@@ -318,9 +318,15 @@ void RenderTreeUpdater::GeneratedContent::updateWritingSuggestionsRenderer(Rende
+
     auto prefix = textWithoutSuggestion.substring(0, offset);
     auto suffix = textWithoutSuggestion.substring(offset);
+#endif
```
I think it would be good to refactor this code, so that we avoid this platform-specific `#if` in WebCore rendering code.

Maybe we could source it from the already-platform-specific venues in `Editor{IOS|Mac}.mm` and then propagate it here, or add a new `EditingBehavior` method.

whsieh commented on Source/WebCore/dom/WritingSuggestionData.h:
```diff
@@ -46,10 +49,20 @@ class WritingSuggestionData {
+
+    String m_originalPrefix;
+    String m_originalSuffix;
+    bool m_supportsSuffix;
```
Nit - might be good to give this an initial value to prevent UB in the future, though I think it's okay either way (up to you!).

Same regarding `m_offset` above

rr-codes committed:
Typing into Facebook post autocompletes too much
https://bugs.webkit.org/show_bug.cgi?id=276206
rdar://129435650

Reviewed by Wenson Hsieh.

Use the new pseudo-style-based writing suggestions implementation on both iOS and macOS, instead of just macOS,
and account for the differences in the behavior between how AppKit and UIKit interface with WebKit for providing
the predictive text.

* Source/WebCore/dom/WritingSuggestionData.h:
(WebCore::WritingSuggestionData::WritingSuggestionData):
(WebCore::WritingSuggestionData::originalPrefix const):
(WebCore::WritingSuggestionData::originalSuffix const):
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::setWritingSuggestion):
(WebCore::Editor::respondToChangedSelection):
* Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
(WebCore::RenderTreeUpdater::GeneratedContent::updateWritingSuggestionsRenderer):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView setAttributedMarkedText:selectedRange:]):

Canonical link: https://commits.webkit.org/281060@main
